spock-core/src/main/java/org/spockframework/runtime/ParameterizedSpecRunner.java

```diff
@@ -112,20 +116,39 @@ private int estimateNumIterations(Object[] dataProviders) {
     return result == Integer.MAX_VALUE ? -1 : result;
   }
 
-  private void runIterations(Iterator[] iterators, int estimatedNumIterations) {
-    if (runStatus != OK) return;
-
-    while (haveNext(iterators)) {
-      initializeAndRunIteration(nextArgs(iterators), estimatedNumIterations);
-
-      if (resetStatus(ITERATION) != OK) break;
+  private void runIterations(final FeatureInfo currentFeature, final Iterator[] iterators, final int estimatedNumIterations) {
+    if (runStatus.get() != OK) return;
+
+    boolean atLeastOneIteration = false;
+
+    while (haveNext(currentFeature, iterators)) {
+      atLeastOneIteration = true;
+      final Object[] dataValues = nextArgs(currentFeature, iterators);
+      
+      if (currentFeature.isReportIterations()){// it is time to schedule independent tasks
```

Comment:
It is due to reporting problems: If !isReportIterations we should notify JUnit listener only once - after all iteration will be finished.